rustls/tests/server_cert_verifier.rs

```diff
@@ -152,3 +163,218 @@ fn client_can_override_certificate_verification_and_offer_no_signature_schemes()
+            .dangerous()
+            .with_custom_certificate_verifier(server_verifier)
+            .with_no_client_auth();
+
+        let (mut client, mut server) =
+            make_pair_for_arc_configs(&Arc::new(cas_unaware_client_config), &server_config);
+
+        cas_unaware_error_count += do_handshake_until_error(&mut client, &mut server)
+            .inspect_err(|e| {
+                assert!(matches!(
+                    e,
+                    ErrorFromPeer::Client(Error::InvalidCertificate(
+                        CertificateError::UnknownIssuer
+                    ))
+                ))
+            })
+            .is_err() as usize;
+
+        println!("key type {key_type:?} success!");
+    }
+
+    // For cas_unaware clients, all of them should fail except one that happens to
+    // have the cert the server sends
+    assert_eq!(cas_unaware_error_count, key_types.len() - 1);
+}
+
+#[derive(Debug, Clone)]
+pub struct ResolvesCertChainByCaName(Vec<(DistinguishedName, Arc<CertifiedKey>)>);
+
+impl ResolvesServerCert for ResolvesCertChainByCaName {
+    fn resolve(&self, client_hello: ClientHello<'_>) -> Option<Arc<CertifiedKey>> {
+        let Some(cas_extension) = client_hello.certificate_authorities() else {
+            println!("ResolvesCertChainByCaName: no CAs extension in ClientHello, returning default cert");
+            return Some(self.0[0].1.clone());
+        };
+        for (name, certified_key) in self.0.iter() {
+            let name = X509Name::from_der(name.as_ref())
+                .unwrap()
+                .1;
+            if cas_extension.iter().any(|ca_name| {
+                X509Name::from_der(ca_name.as_ref()).is_ok_and(|(_, ca_name)| ca_name == name)
+            }) {
+                println!("ResolvesCertChainByCaName: found matching CA name: {name}");
+                return Some(certified_key.clone());
+            }
+        }
+        println!("ResolvesCertChainByCaName: no matching CA name found, returning default Cert");
+        Some(self.0[0].1.clone())
+    }
+}
+
+#[derive(Debug)]
+struct ServerCertVerifierWithCasExt {
+    verifier: Arc<dyn ServerCertVerifier>,
+    ca_names: Vec<DistinguishedName>,
+}
+
+impl ServerCertVerifier for ServerCertVerifierWithCasExt {
+    fn verify_server_cert(
+        &self,
+        end_entity: &pki_types::CertificateDer<'_>,
+        intermediates: &[pki_types::CertificateDer<'_>],
+        server_name: &pki_types::ServerName<'_>,
+        ocsp_response: &[u8],
+        now: pki_types::UnixTime,
+    ) -> Result<rustls::client::danger::ServerCertVerified, Error> {
+        self.verifier
+            .verify_server_cert(end_entity, intermediates, server_name, ocsp_response, now)
+    }
+
+    fn verify_tls12_signature(
+        &self,
+        message: &[u8],
+        cert: &pki_types::CertificateDer<'_>,
+        dss: &rustls::DigitallySignedStruct,
+    ) -> Result<rustls::client::danger::HandshakeSignatureValid, Error> {
+        self.verifier
```

Comment:
Suggest importing all the qualified types here.

Reply:
Fixed!